src/mpi/coll/src/csel.c

```diff
@@ -615,63 +615,6 @@ static csel_node_s *prune_tree(csel_node_s * root, MPIR_Comm * comm_ptr)
                     node = node->failure;
                 break;
 
-            case CSEL_NODE_TYPE__OPERATOR__COMM_SIZE_LE:
-                if (comm_ptr->local_size <= node->u.comm_size_le.val)
-                    node = node->success;
-                else
-                    node = node->failure;
-                break;
```

Comment:
I didn't notice these modifications to CSEL in my initial review. Are they necessary? Since CSEL is co-developed with Intel, we're going to need a review of this from @pmodels/intel-reviewers before it can be accepted.

Reply:
This is added after your initial review. It is a proper patch replacing the previous temp commits. They are necessary if we use CSEL in sub-collectives, i.e. collectives with subgroup. Because the size of subgroup is different from the original comm_size, obviously the pruning based on comm_size will be wrong for subgroup collectives.

This only impacts the csel trees that the top tree node is a comm size comparison, and it only saves a single tree descendence. So I don't think there will be any measurable performance gain from pruning, and we have to ensure correctness.

Reply:
> This only impacts the csel trees that the top tree node is a comm size comparison

I'm not familiar enough with CSEL to know how common these tree types are, so I'm not sure how to make a judgement.

Reply:
If subgroup is to replace sub-communicator, it might deserve its own pruned CSEL tree. I think having tree pruned as much as possible is needed for selection performance. Comm size selection knob is often at the very top of the Json tree, not pruning it probably means all communicator is using the full CSEL tree.

Reply:
You can assume every collective selection will first check for comm_size, that is one check you can save if you prune it away. I believe the saving is insignificant compared to the infrastructure of CSEL itself -- even the NULL selection overhead will shadow the pruning saving. Anyway, the pruning on comm size or hierarchy will be incorrect for subgroup collectives. We certainly can get the pruning back. One option is to prune it at group level. Another option is to apply simple branch prediction. Both IMO are premature optimizations.

Reply:
But this PR is changing *existing behavior*. That is the source of my hesitation. Saying it doesn't actually matter and we can fix it later is not reassuring.

Reply:
The issue is that it may stop the pruning prematurely at comm_size level which results in a full tree with virtually no pruning. I think the first option to prune it at group level seems to be consistent with the old behavior.

Reply:
Benchmarking 4-process intranode bcast (more processes or adding internode will only increase the latency thus further hide the pruning difference) -
#### With Pruning 
     msgsize      min      max      avg      sigma  (in microseconds)
           0    0.025    0.026    0.025      0.050
           1    0.208    0.265    0.236      0.239
           2    0.209    0.263    0.235      0.145
           4    0.213    0.260    0.234      0.155
           8    0.210    0.259    0.233      0.025
          16    0.207    0.264    0.235      0.240
          32    0.207    0.266    0.235      0.188
          64    0.206    0.263    0.234      0.020
         128    0.212    0.290    0.254      0.040
         256    0.211    0.322    0.281      0.140
         512    0.217    0.323    0.293      0.167
        1024    0.228    0.501    0.430      0.169
        2048    0.233    0.644    0.539      0.043
        4096    0.357    0.894    0.759      0.054
        8192    0.589    1.492    1.257      0.054
       16384    1.307    2.352    2.083      0.502
       32768    2.307    4.063    3.618      0.997
       65536    4.379    6.261    5.776      0.833
      131072    9.789   11.760   11.262      1.031
      262144   21.601   23.590   23.083      0.387
      524288   42.032   44.034   43.522      1.965
     1048576   90.758   92.669   92.174      3.790
     2097152  188.395  190.482  189.945      4.543
     4194304  394.832  396.833  396.302      7.462
#### Without pruning
     msgsize      min      max      avg      sigma  (in microseconds)
           0    0.025    0.026    0.025      0.053
           1    0.213    0.271    0.240      0.259
           2    0.215    0.270    0.241      0.218
           4    0.221    0.261    0.238      0.131
           8    0.213    0.267    0.239      0.223
          16    0.214    0.266    0.238      0.179
          32    0.214    0.262    0.237      0.033
          64    0.213    0.264    0.237      0.026
         128    0.221    0.299    0.260      0.100
         256    0.225    0.340    0.297      0.210
         512    0.225    0.309    0.284      0.099
        1024    0.246    0.520    0.451      0.216
        2048    0.236    0.629    0.529      0.046
        4096    0.361    0.895    0.760      0.044
        8192    0.663    1.553    1.328      0.053
       16384    1.160    2.330    2.035      0.079
       32768    2.395    4.104    3.664      1.020
       65536    4.831    6.644    6.186      0.828
      131072    9.905   11.863   11.369      1.043
      262144   20.779   22.776   22.272      1.167
      524288   41.091   43.201   42.666      2.151
     1048576   89.024   91.075   90.535      3.210
     2097152  196.788  198.947  198.405      5.254
     4194304  408.871  410.957  410.297      8.929
#### Conclusion
Looking at the small message numbers where latency are most sensitive, one may see ~0.01 us performance benefit from pruning. This is way below the noise and insignificant IMO.

Reply:
This is a too small sample. Not enough to draw any conclusion. We can probably do a thorough study on pruning in terms of performance and memory usage, and we may even decide to totally remove pruning once for all if that is the case. It could be a separate PR.

Reply:
The pruning effect is entirely a local instruction benefit common to all collectives, so I don't see how the scale matters. Any scale will only further hide any pruning benefit. Unfortunately, it can not be a separate PR because in this PR it is a correctness issue.

Reply:
Why can't we implement pruning for subgroup same as sub-comm? I mean since we are still doing pruning in this PR, we should then do pruning properly like we did for sub-communicator. If for correctness, and literally nullify the pruning, then that changes the existing behavior and it should be done in a separate PR and potentially we can just remove pruning once for all, instead of having a half functioning pruning.

Reply:
The whole point of subgroup is lightweight and flexibility. Yes, we can do caches and init-optimizations at subgroup level, but I consider them premature optimization and will hinder future development. 

I think it's a good idea to remove the CSEL pruning. I can make a separate PR removing the csel pruning if you agree.

Reply:
My opinion is still the same. Caching scheme sounds a right solution for me. I don't know why it is considered premature optimization. This PR should not change the pruning behavior for common scenarios (i.e. not using subgroups), to decide whether to remove pruning belongs to a new PR.
Unfortunately people who originally worked on pruning have all left and we lost some institutional knowledge of pruning and its benefits. Other people @pmodels/intel-reviewers can chime in.